tests/scripts/all.sh

```diff
@@ -4391,6 +4391,63 @@ component_build_aes_aesce_armcc () {
 }
 
+support_build_aes_armce() {
+    # clang >= 4 is required to build with AES extensions
+    ver="$(clang --version|grep version|sed -E 's#.*version ([0-9]+).*#\1#')"
+    [ "${ver}" -ge 11 ]
+}
+
+component_build_aes_armce () {
+    # Test variations of AES with Armv8 crypto extensions
+    scripts/config.py set MBEDTLS_AESCE_C
+    scripts/config.py set MBEDTLS_AES_USE_HARDWARE_ONLY
+
+    msg "MBEDTLS_AES_USE_HARDWARE_ONLY, clang, aarch64"
+    make -B library/aesce.o CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a+crypto"
+
+    msg "MBEDTLS_AES_USE_HARDWARE_ONLY, clang, arm"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a72+crypto -marm"
+
+    msg "MBEDTLS_AES_USE_HARDWARE_ONLY, clang, thumb"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a32+crypto -mthumb"
+
+    scripts/config.py unset MBEDTLS_AES_USE_HARDWARE_ONLY
+
+    msg "no MBEDTLS_AES_USE_HARDWARE_ONLY, clang, aarch64"
+    make -B library/aesce.o CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a+crypto"
+
+    msg "no MBEDTLS_AES_USE_HARDWARE_ONLY, clang, arm"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a72+crypto -marm"
+
+    msg "no MBEDTLS_AES_USE_HARDWARE_ONLY, clang, thumb"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a32+crypto -mthumb"
+
+    # test for presence of AES instructions
+    scripts/config.py set MBEDTLS_AES_USE_HARDWARE_ONLY
+    msg "clang, test A32 crypto instructions built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a72+crypto -marm -S"
+    grep -E 'aes[0-9a-z]+.[0-9]\s*[qv]' library/aesce.o
+    msg "clang, test T32 crypto instructions built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a32+crypto -mthumb -S"
+    grep -E 'aes[0-9a-z]+.[0-9]\s*[qv]' library/aesce.o
+    msg "clang, test aarch64 crypto instructions built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a -S"
+    grep -E 'aes[a-z]+\s*[qv]' library/aesce.o
+
+    # test for absence of AES instructions
+    scripts/config.py unset MBEDTLS_AES_USE_HARDWARE_ONLY
+    scripts/config.py unset MBEDTLS_AESCE_C
+    msg "clang, test A32 crypto instructions not built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a72+crypto -marm -S"
+    not grep -E 'aes[0-9a-z]+.[0-9]\s*[qv]' library/aesce.o
+    msg "clang, test T32 crypto instructions not built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=arm-linux-gnueabihf -mcpu=cortex-a32+crypto -mthumb -S"
+    not grep -E 'aes[0-9a-z]+.[0-9]\s*[qv]' library/aesce.o
+    msg "clang, test aarch64 crypto instructions not built"
+    make -B library/aesce.o CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a -S"
```

Comment:
@tom-cosgrove-arm I'm confused as to exactly in what setting we test the absence of AES instructions. Specifically, why are we testing that on `armv8-a` and not `armv8-a+crypto`? Is that just copypasta or is there a reason?

Reply:
Actually, I think the errors are on lines 4441 and 4444. Here we are checking that we don't emit AES instructions when we don't ask for them, so that (for example) we won't get SIGILL when run on an older Raspberry Pi (IIRC the RPi5 is the first with crypto extensions). If we passed `+crypto` to the compiler it would be allowed to use any of those instructions itself, so could give false positives. So not having `+crypto` here is correct, but having `+crypto` in the two cases above is incorrect

Reply:
I see, thanks! I'll fix it in https://github.com/Mbed-TLS/mbedtls/pull/9286.

Hmmm...

> we don't emit AES instructions when we don't ask for them

But if we're telling the compiler not to use AES instructions, how would the test ever fail? If we were enabling assembly that calls AES instructions?

Reply:
And below with sha256:

```
            make -B library/sha256.s CC=clang CFLAGS="--target=aarch64-linux-gnu -march=armv8-a"
            grep -E 'sha256[a-z0-9]+\s+[qv]' library/sha256.s
```

So we do find sha256 instructions on `armv8-a`, presumably from our assembly?

Reply:
Just following this thread for interest, but I just noticed and was about to raise: does this grepping work? It seems like we are grepping for the assembler mnemonic in the object file, which doesn't seem like it would work, even if we have intrinsics.

Surely we need to `objdump` it first?

Reply:
The `-S` in `CFLAGS` means "output assembly, not object code, into the output file" (which we still name with`.o`, but hey-ho)

Reply:
Ah okay, makes sense, thanks!

Reply:
```
commit acbdbc7c2b32368b07a5b49a170b202d880afa12
Author: Gilles Peskine <Gilles.Peskine@arm.com>
Date:   2024-06-19 14:16:05 +0200

    Use .s extension for assembly
    
    Having assembly files called *.o was confusing.
```